### PR TITLE
Support getting history of deployments for a service and listing running services (or any given status)

### PR DESCRIPTION
**Pull request by indiejames**
## Jira Issue ID
HARMONY-1749/1750

## Description
Adds support for listing deployments that have been done for a particular service as well as listing deployments by status,
e.g., got get the running deployment.

## Local Test Steps
Do a service  deployment
```
curl -Ln -bj -XPUT "http://localhost:3000/service-image-tag/query-cmr" -H 'Authorization: bearer <your token>' -d '{"tag": "latest"}' -H 'Content-type: application/json'
```
List all service deployments
```
curl -Ln -bj "http://localhost:3000/service-image-tag/deployment"
```
List service deployments for query-cmr
```
curl -Ln -bj "http://localhost:3000/service-image-tag/deployment?service-id=query-cmr"
```
List service deployments by status
```
curl -Ln -bj "http://localhost:3000/service-image-tag/deployment?status=failed"
```

## PR Acceptance Checklist
* [x] Acceptance criteria met
* [x] Tests added/updated (if needed) and passing
* [x] Documentation updated (if needed)

### CODE REVIEW

File: services/harmony/app/frontends/service-image-tags.ts [hunk 1]
```diff
@@ -501,5 +513,35 @@ export async function getServiceDeployment(
...
+ */
+export async function getServiceDeployments(
+  req: HarmonyRequest, res: Response, next: NextFunction,
+): Promise<void> {
+  const validations = [
+    validateUserIsInDeployerOrAdminGroup,
+    validateStatusForListServiceRequest,
+  ];
+
+  for (const validation of validations) {
```
[the remark this answers is not in the document]
Comment (indiejames) [reply services/harmony/app/frontends/service-image-tags.ts:533]: I'll add a coment explaing that 'no status' is a valid choice.

File: services/harmony/app/frontends/service-image-tags.ts [hunk 2]
```diff
@@ -501,5 +513,35 @@ export async function getServiceDeployment(
...
+    if (! await validation(req, res)) return;
+  }
+  try {
+    await db.transaction(async (tx) => {
+      const queryLowerCase = keysToLowerCase(req.query);
+      const deployments = await getDeployments(tx, queryLowerCase.status, queryLowerCase.serviceid);
```
[the remark this answers is not in the document]
Comment (indiejames) [reply services/harmony/app/frontends/service-image-tags.ts:539]: Vinny's fault :). I agree, `service` would be better. Done.

File: services/harmony/app/models/service-deployment.ts [hunk 3]
```diff
@@ -2,6 +2,16 @@ import { Transaction } from './../util/db';
...
 import { truncateString } from '@harmony/util/string';
 
+export enum ServiceDeploymentStatus {
+  RUNNING = 'running',
+  SUCCESSFUL = 'successful',
+  FAILED = 'failed',
+}
+
+export type ServiceDeploymentForDisplay =
+  Omit<ServiceDeployment, 'deployment_id' | 'id' | 'validate' | 'save' | 'serialize'> &
```
[the remark this answers is not in the document]
Comment (indiejames) [reply services/harmony/app/models/service-deployment.ts:12]: Ask Vinny?

File: services/harmony/app/frontends/service-image-tags.ts [hunk 4]
```diff
@@ -274,6 +275,28 @@ async function validateTagPresent(
...
+ * Verify that the requested service deployment status is one of the valid statuses
+ * @param req - The request object
+ * @param res  - The response object - will be used to send an error if the validation fails
+ * @returns a Promise containing `true` if the status is valid, `false` otherwise
+ */
+async function validateStatusForListServiceRequest(
+  req: HarmonyRequest, res: Response,
+): Promise<boolean> {
+  const { status } = req.query;
+  if (status) {
```
[the remark this answers is not in the document]
Comment (indiejames) [reply services/harmony/app/frontends/service-image-tags.ts:290]: I added a comment that passing no status was a valid choice - we only validate 'status' if it is present.

File: services/harmony/app/frontends/service-image-tags.ts [hunk 5]
```diff
@@ -274,6 +275,28 @@ async function validateTagPresent(
...
+    const validStatuses = Object.values(ServiceDeploymentStatus).map((val) => JSON.stringify(val));
```
[the remark this answers is not in the document]
Comment (indiejames) [reply services/harmony/app/frontends/service-image-tags.ts:292]: I add case-insensitiviy to the 'status' check

[the remark this answers is not in the document]
Comment (chris-durbin) [reply services/harmony/app/frontends/service-image-tags.ts:292]: There needs to be a sad emoji for my disappointment in TypeScript.

File: docs/guides/managing-existing-services.md
Comment (ygliuvt) [inline docs/guides/managing-existing-services.md:1]: We should update the documentation on `GET /service-deployments-state` endpoint.

**Review by chris-durbin — APPROVED** [review]

**Review by ygliuvt — CHANGES_REQUESTED** [review]
Failed to test locally, getting the following error with `service-deployments-state` PUT requests:
```
User undefined is not in the admin EDL group
```
The issue is caused by the new service-deployment* routes are not in the `authorizedRoutes` list. We can add `'/service-deployment*',` in the `authorizedRoutes` list to fix the issue.

File: docs/guides/managing-existing-services.md
Comment (indiejames) [reply docs/guides/managing-existing-services.md:1]: Done.

**Comment by indiejames** [comment]
> Failed to test locally, getting the following error with `service-deployments-state` PUT requests:
> 
> ```
> User undefined is not in the admin EDL group
> ```
> 
> The issue is caused by the new service-deployment* routes are not in the `authorizedRoutes` list. We can add `'/service-deployment*',` in the `authorizedRoutes` list to fix the issue.

Good catch. I added '/service-deployment*' to `authorizedRoutes`.

**Review by ygliuvt — CHANGES_REQUESTED** [review]
I found query with both `service` and `status` not working. e.g. when I test locally the following query: `http://localhost:3000/service-deployment?service=giovanni-adapter&status=successful` returns the only harmony-service-example service deployment that is successful. We can support this a separate ticket.

**Review by ygliuvt — APPROVED** [review]

**Pull request merged**